src/debug_sdfs_64drive.c

```diff
@@ -74,11 +74,28 @@ static DRESULT fat_disk_write_64drive(const BYTE* buff, LBA_t sector, UINT count
 
 			uint32_t* dst = (uint32_t*)(D64_CIBASE_ADDRESS + D64_BUFFER);
 			u_uint32_t* src = (u_uint32_t*)buff;
+			disable_interrupts();
 			for (int i = 0; i < 512/16; i++)
 			{
 				uint32_t a = *src++; uint32_t b = *src++; uint32_t c = *src++; uint32_t d = *src++;
-				*dst++ = a;          *dst++ = b;          *dst++ = c;          *dst++ = d;
+				dma_wait();
+				MEMORY_BARRIER();
+				*dst++ = a;
+				MEMORY_BARRIER();
+				dma_wait();
+				MEMORY_BARRIER();
+				*dst++ = b;
+				MEMORY_BARRIER();
+				dma_wait();
+				MEMORY_BARRIER();
+				*dst++ = c;
+				MEMORY_BARRIER();
+				dma_wait();
+				MEMORY_BARRIER();
+				*dst++ = d;
+				MEMORY_BARRIER();
 			}
+			enable_interrupts();
```

Comment:
I'm a bit worried on the interrupt latency here. Can you please measure how long we are keeping the interrupts disabled here? Also `dma_wait()` seems wrong because of its name (though I know it's doing what we expect). Maybe we should simply switch to `io_write()` / `__io_write()` here (the second if we keep the interrupts disabled ourselves)?

So I think it would be better to measure with TICKS_READ() how long it takes to do a whole run in the two scenarios: 1) keep interrupt disabled, use `__io_write()`, 2) keep interrupts enabled, call `io_write()`. The second version would switch interrupts on/off between each word write. A third option would be to enable/disable within the for cycle so that at least we re-enable them once per loop.

Reply:
I was thinking that to generalize this, we could have an API `pi_lock()` / `pi_unlock()` to lock/unlock the bus, where locking does the loop dance above to try to acquire the lock without keeping interrupts disabled and pi_unlock() just reenables the interrupts. So all users of the PI bus have a simple API to handle it.

Reply:
Once we acquire the lock, if we have a long running code, it will still introduce latency if we just use interrupts for the lock. Maybe we should work on an alternative lockless solution. (Not that I have in mind ATM)
On the other hand I totally get your point we should clarify the intent of the API (I know dma_wait was not very appropriate but wanted to start discussions around). Once we do that the locking itself can be an implementation detail though it would still matter for this PR as it would change the implementation here.
Also for very small read/writes like we have in this PR we wouldn't want too much overhead.
Then if we utilize `C11` threads, I think we can implement safer and more performant threading, WDYT? The compiler can cooperate with our needs as well that way.

Reply:
Yes the idea of pi_lock is to help programmers acquire the lock in the correct way. Right now, exactly zero code does that (well, just the RDP), so it looks like the pattern is easy to be mistaken.

pi_lock doesn't prevent us from the fact that, once a lock is taken, it must be retained for not too long. I don't think there's an easy way. We currently hide the lock pattern behind high level primitive like io_read but that's a bit too little (just one memory access). I can't think of a way to design a API that allows for some more packing of memory accesses but not too many.
It's also hard to differentiate standard cases from special cases as PI is just a bus. Maybe sometimes you really need to keep bus locked and ignore interrupt latency, so it would be wrong to design a API that handcuffs you in this regard.


C11 threads is just an API. We could decide to implement it on top of our kernel if we wanted, but it's the implementation that matters. In our case, I don't think we can allow anything to happen while the lock is held: almost anything could touch the PI bus. So the only way is to have the PI lock keep the interrupts disabled.

Reply:
It should be possible to have interrupts fire and execute as long as they do not touch PI IINM. Once they do, they can start waiting for the release of the lock.

In theory it can be more than an API. I really don't know otimizations in GCC or the memory model details of C11 but in theory it can better (and more correctly) optimize code if it is thread aware. It is not possible if we don't cooperate with the compiler.

See http://www.open-std.org/jtc1/sc22/wg21/docs/papers/2015/n4455.html for some possible optimization examples.

Reply:
> It should be possible to have interrupts fire and execute as long as they do not touch PI IINM. 

Which is not possible to enforce, and, what's worse, it's a problem only in the rare case where the interrupt is fired with the exact timing within the lock, so it will cause hard to reproduce and debug bugs.

> Once they do, they can start waiting for the release of the lock.

If the interrupt wait for the lock to be released, it will deadlock because the lock is held in user space, which will not resume until the interrupt is finished. If instead the interrupt bypasses the lock but just waiting for PI ready, when the user space resume execution the lock would have been violated and maybe the PI is now busy because of a DMA started under interrupt... but the user space thinks it has acquired the lock.

Reply:
Ok I was incorrectly referring to `dma_wait` by "wait" here. `pi_lock` is the necessary devil for such small one off read things then? IDK what would be best b/c with such a lock method, there would be too much overhead.